### PR TITLE
Add min_entropy function to quantum information module

### PR DESCRIPTION
**Pull request by EuGig**
------------------------------------------------------------------------------------------------------------

**Context**:
Add min_entropy function to quantum information module

**Description of the Change**:
Addiction of the min_entropy function in pennylane/math/quantum.py that takes either a state vector or a density matrix and logarithm base as input, and outputs its minimum entropy.

**Benefits:**
It enables to calculate the min-entropy.

**Possible Drawbacks:**

**Related GitHub Issues:**

Fixes https://github.com/PennyLaneAI/pennylane/issues/3141

### CODE REVIEW

**Comment by EuGig** [comment]
What is wrong with the formatting check? I honestly don't know much about black formatting.

**Comment by EuGig** [comment]
What is wrong with the formatting check? I honestly don't know much about black.

**Comment by trbromley** [comment]
Thanks @EuGig for the pull request! We'll take a look soon and give some feedback.

**Review by rmoyard — COMMENTED** [review]
Thanks for the PR @EuGig!

I have a couple of comments. The min entropy cannot be negative, so it is missing a minus sign and it would be great to test it for a least one value where it is not equal to the max and vn entropy. https://en.wikipedia.org/wiki/R%C3%A9nyi_entropy#Min-entropy

Let me know if you have any questions!

File: pennylane/math/quantum.py
Comment (EuGig) [reply pennylane/math/quantum.py:1154]: Oh, that's why I messed up files while computing from my terminal. I'm also adding tests for values different from max and vn entropy.

Comment (EuGig) [reply pennylane/math/quantum.py:1154]: I was thinking to write a test for the following state: [1, 0, 1, 0] / np.srt(2), and I got the following values:
min-entropy: 2.2204460492503136e-16
Vn-entropy: 2.2204460492503136e-16
Max-entropy: 0.0
And I don't understand how the min and vn-entropy are higher than the max-entropy. Should'n they be always less or equal?

File: pennylane/math/quantum.py
Comment (albi3ro) [inline pennylane/math/quantum.py:1206]:
```suggestion
   div_base = np.log(base) if base else 1
```
Little tidier.

File: pennylane/math/quantum.py
Comment (albi3ro) [inline pennylane/math/quantum.py:1178]:
```suggestion
    return _compute_min_entropy(density_matrix, base)
```

This isn't in pylint or black, but some other linters recommend "inline variable that is immediately returned". This is just a personal preference thing, so feel free to ignore the comment.

File: pennylane/math/quantum.py
Comment (EuGig) [reply pennylane/math/quantum.py:1154]: Thank you!! @albi3ro. I added the state [1,0,1,0]/np.srt(2) to testing as @rmoyard suggested. However, it's not included into differentiability tests.

File: pennylane/math/quantum.py
Comment (trbromley) [inline pennylane/math/quantum.py:1132]:
```suggestion
    r"""Compute the minimum entropy from a state vector or density matrix on a given subsystem.
```

File: pennylane/math/quantum.py
Comment (trbromley) [inline pennylane/math/quantum.py:1152]:
```suggestion
    >>> x = [1, 0, 0, 1] / np.sqrt(2)
    >>> x = dm_from_state_vector(x)
```

File: pennylane/math/quantum.py
Comment (trbromley) [inline pennylane/math/quantum.py:1138]: Can we make this accept just density matrices? See #4322.

File: pennylane/math/quantum.py
Comment (trbromley) [inline pennylane/math/quantum.py:1169]:
```suggestion
    >>> x = [np.cos(np.pi/8), 0, 0, -1j*np.sin(np.pi/8)]
    >>> x = dm_from_state_vector(x)
```

File: pennylane/math/quantum.py
Comment (trbromley) [inline pennylane/math/quantum.py:1201]:
```suggestion
```

File: tests/math/test_entropies_math.py
Comment (trbromley) [inline tests/math/test_entropies_math.py:451]:
```suggestion

```

File: pennylane/math/__init__.py
Comment (trbromley) [inline pennylane/math/__init__.py:77]: Also need to add to `__all__` (further down in this file)

File: pennylane/math/quantum.py
Comment (trbromley) [reply pennylane/math/quantum.py:1138]: The code itself might not have to change much, but it would be good to add a test for batching.

**Review by trbromley — CHANGES_REQUESTED** [review]
Thanks @EuGig! This looks great and apologies for the delay in getting back to this PR. I've left a few suggestions, and we also need to put an entry in the changelog+optionally your name in the list of contributors. Other than that, we should be good to merge soon.

File: pennylane/math/quantum.py
Comment (EuGig) [reply pennylane/math/quantum.py:1027]: Hi @trbromley and thank you for the review! Do you have any suggestion about how to use batching with density matrices? Tests for similar functions in `quantum.py` don't seem to use it; not even `test_density_matrices_max_entropy` (A similar test for the min-entropy is already present).

File: tests/math/test_entropies_math.py
Comment (timmysilv) [inline tests/math/test_entropies_math.py:377]: otherwise, this test will do nothing when `interface` is `None`
```suggestion
        entropy = qml.math.min_entropy(state_vector, wires, base, check_state=check_state)

        if pure:
            expected_min_entropy = 0
        else:
            expected_min_entropy = np.log(2) / np.log(base)
        assert qml.math.allclose(entropy, expected_min_entropy)
```

**Review by timmysilv — COMMENTED** [review]
Hi @EuGig 👋  excited to have this in PennyLane! There are still a few things left to get done, but then we should be good to go.

I appreciate you bearing with the team and addressing review comments. I see that you've marked a handful of them as "resolved", but the changes haven't yet been implemented. If you've got them locally, make sure to push them. Otherwise, we'll be ready when you are! I have un-resolved the conversations that still need attention.

Also, it seems that your recent merge of PennyLane master into your fork had some conflicts, so there are a few random "master" instances in the code - if you remove those, the formatting/docs/pylint CI should hopefully be nice and turn green.

Lastly, Tom mentioned PR #4322 above, so let me add a bit of context and clarify what's happened. When you started on this PR, most of the functions in `quantum.py` accepted either a statevector _or_ a density matrix. The linked PR changed this standard, and now those functions (including `min_entropy` 🎉 ) should not accept statevectors - only density matrices. This will require a few changes from your end, namely:
- update the function parameters and docs to reflect this change
- remove duplicate tests that use statevectors (we only need to test that `min_entropy` works with statevectors, as the function `dm_from_state_vector` is already tested elsewhere)

Don't worry too much about the batching, but if you're curious, you can look at the tests in the `TestEntropyBroadcasting` class below. We use the terms "batching" and "broadcasting" interchangeably sometimes, although they are not exactly the same. Sorry for the confusion! I think broadcasting support will require some additional work, so perhaps we can save it for a follow-up PR 😄 Let me know if I can help with any of these things 💪

File: pennylane/math/quantum.py
Comment (timmysilv) [reply pennylane/math/quantum.py:1027]: added some details in my latest comment below.

**Comment by EuGig** [comment]
Hi @timmysilv!! thank you for the thorough comment. Even though I run `pylint --rcfile tests/.pylintrc tests/math/test_entropies_math.py` as you suggested, pylint is still compelling about this three global variables:
```
24 tf = pytest.importorskip("tensorflow", minversion="2.1")
25 torch = pytest.importorskip("torch")
26 jax = pytest.importorskip("jax")
```

The problem is that the import is locally (inside test classes) repeated. So, I simply deleted the local imports.
Besides this, Jax tests are doing something weird. Some of them got cancelled.

**Comment by timmysilv** [comment]
hey again, it's a little hard to read the relevant changes along with all the auto-generated changes from `black` - they released a new version and we didn't expect it, sorry about that! I'm not able to make changes to your fork, but I made a branch on the main repo called `timmysilv-patch` that resolves all the conflicts. you can run the following to resolve them on this PR too:
1. `git fetch origin`
    * if that failed, run `git remote add origin https://github.com/PennyLaneAI/pennylane.git`
3. `git merge origin/timmysilv-patch`
4. `git push`

The good news is, the [diff looks good](https://github.com/PennyLaneAI/pennylane/compare/b5f302c5fc903762ea083b5f7df19148cc06c23b...timmysilv-patch) from that patch, so I was able to confirm that your changes are nearly good to go! The only thing I'm seeing is an incorrect indent on the 2 times in the docstring where you wrote "`>>> x = dm_from_state_vector(x)`". After merging in my changes and fixing the indent, I'd say this PR is good to go!

File: pennylane/math/quantum.py
Comment (timmysilv) [inline pennylane/math/quantum.py:1041]:
```suggestion
    >>> x = dm_from_state_vector(x)
```

File: pennylane/math/quantum.py
Comment (timmysilv) [inline pennylane/math/quantum.py:1059]:
```suggestion
    >>> x = dm_from_state_vector(x)
```

**Comment by EuGig** [comment]
After some struggle I think I managed to do the right changes! :sweat_smile:

File: pennylane/math/quantum.py
Comment (timmysilv) [inline pennylane/math/quantum.py:1059]:
```suggestion
    >>> x = dm_from_state_vector(x)
```

File: pennylane/math/quantum.py
Comment (timmysilv) [inline pennylane/math/quantum.py:1041]:
```suggestion
    >>> x = dm_from_state_vector(x)
```

**Review by timmysilv — COMMENTED** [review]
sorry about that, I had merged in the fixes already and I think it caused some strangeness. I'll fix it and approve once CI passes!

File: tests/math/test_entropies_math.py
Comment (astralcai) [inline tests/math/test_entropies_math.py:473]: I'm probably just nitpicking here but this section of code is duplicated 4 times.

**Review by astralcai — COMMENTED** [review]
Looks great!

File: tests/math/test_entropies_math.py
Comment (EuGig) [reply tests/math/test_entropies_math.py:473]: Actually, I checked and the section you mentioned doesn't seem necessary, considering the kind of parameters used.

File: tests/math/test_entropies_math.py
Comment (timmysilv) [inline tests/math/test_entropies_math.py:369]:
```suggestion
        assert qml.math.allclose(gradient, -np.eye(4) / np.log(base))
```

File: tests/math/test_entropies_math.py
Comment (timmysilv) [inline tests/math/test_entropies_math.py:385]:
```suggestion
        assert qml.math.allclose(gradient, -np.eye(4) / np.log(base))
```

File: tests/math/test_entropies_math.py
Comment (timmysilv) [inline tests/math/test_entropies_math.py:402]:
```suggestion
        assert qml.math.allclose(gradient, -np.eye(4) / np.log(base))
```

File: tests/math/test_entropies_math.py
Comment (timmysilv) [inline tests/math/test_entropies_math.py:428]:
```suggestion
        assert qml.math.allclose(gradient, -np.eye(4) / np.log(base))
```

**Review by timmysilv — COMMENTED** [review]
i think you had it right before 😄

**Review by astralcai — APPROVED** [review]
Looks good, thank you for contributing!

**Comment by EuGig** [comment]
Thank you all for the support!!

**Comment by trbromley** [comment]
Thanks @EuGig for contributing! This PR should make it out as part of our 0.35 release on March 5th.

We often tag contributors in our release marketing on Twitter/X. If you would optionally like to be tagged, please could you share your username?

**Comment by EuGig** [comment]
Yes sure!

> @gigante_eu11622

Thank you!

**Pull request merged**
